bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,193 @@
+                    if backend_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    // TODO: replace this panic with an error
+                    panic!("Upstream not found");
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // Set SNI
+        let mut http_peer = HttpPeer::new(upstream, false, "".to_string());
+        let peer_opts = http_peer.get_mut_peer_options().unwrap();
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = TIMEOUT_SECS;
+        peer_opts.read_timeout = TIMEOUT_SECS;
+        peer_opts.write_timeout = TIMEOUT_SECS;
+        peer_opts.idle_timeout = TIMEOUT_SECS;
+
+        peer_opts.alpn = ALPN::H2;
+
+        let peer = Box::new(http_peer);
+        Ok(peer)
+    }
+
+    async fn upstream_request_filter(
+        &self,
+        _session: &mut Session,
+        upstream_request: &mut RequestHeader,
+        _ctx: &mut Self::CTX,
+    ) -> Result<()>
+    where
+        Self::CTX: Send + Sync,
+    {
+        // Check if it's a gRPC request
+        if let Some(content_type) = upstream_request.headers.get("content-type") {
+            if content_type == "application/grpc" {
+                // Ensure the correct host and gRPC headers are forwarded
+                // upstream_request.insert_header("Host", "0.0.0.0").unwrap();
+                upstream_request.insert_header("content-type", "application/grpc").unwrap();
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn request_filter(&self, session: &mut Session, _ctx: &mut Self::CTX) -> Result<bool>
+    where
+        Self::CTX: Send + Sync,
+    {
+        let client_addr = session.client_addr();
+        let user_id = client_addr.map(|addr| addr.to_string());
+
+        // Request ID is a random number
+        let request_id = rand::random::<u64>().to_string();
+        session
+            .req_header_mut()
+            .insert_header("X-Request-ID", request_id)
+            .expect("Failed to insert header");
+
+        // retrieve the current window requests
+        let curr_window_requests = RATE_LIMITER.observe(&user_id, 1);
+
+        if curr_window_requests > MAX_REQ_PER_SEC {
+            // rate limited, return 429
+            let mut header = ResponseHeader::build(429, None).unwrap();
+            header.insert_header("X-Rate-Limit-Limit", MAX_REQ_PER_SEC.to_string()).unwrap();
+            header.insert_header("X-Rate-Limit-Remaining", "0").unwrap();
+            header.insert_header("X-Rate-Limit-Reset", "1").unwrap();
+            session.set_keepalive(None);
+            session.write_response_header(Box::new(header), true).await?;
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    async fn logging(&self, session: &mut Session, e: Option<&Error>, _ctx: &mut Self::CTX)
+    where
```

Comment:
In my first iteration I tried with that particular method but end up realizing that if chose to use `response_filter` I was going to need to replicate that dequeueing logic in the `fail_to_proxy` method too. I ended up using `logging` because is a (in fact, the only) common place between successful and fail requests.